src/client.ts

```diff
@@ -175,22 +159,37 @@ export async function sendGet<
 		return Promise.reject(queryParams.error)
 	}
 
-	return wretch
-		.get(`${params.path}${queryParams.result}`)
-		.json()
-		.then((response) => {
-			const parsedResponse = parseResponseBody({
-				response: response as ResponseBody,
-				responseBodySchema: params.responseBodySchema,
-				path: params.path,
-			})
-
-			if (isFailure(parsedResponse)) {
-				return Promise.reject(parsedResponse.error)
+	return wretch.get(`${params.path}${queryParams.result}`).res(async (response) => {
+		const bodyParseResult = await tryToResolveJsonBody(
+			response,
+			params.path,
+			params.responseBodySchema,
+		)
+
+		if (bodyParseResult.error === 'NOT_JSON') {
+			if (params.isNonJSONResponseExpected) {
+				return response as unknown as Promise<ResponseBody>
+			}
+			return Promise.reject(
+				`Request to ${params.path} has returned an unexpected non-JSON response.`,
+			)
+		}
+
+		if (bodyParseResult.error === 'EMPTY_RESPONSE') {
+			if (params.isEmptyResponseExpected) {
+				return response as unknown as Promise<ResponseBody>
```

Comment:
@ondrejsevcik @szymonchudy How would FE prefer to be communicated about special cases? Currently it would just return entire Response object in case something isn't considered a direct error, but is an unusual situation (e. g. non-JSON or expected as possible empty content came back). Is that good enough? We can return an Either with a defined set of possible result codes instead.

Reply:
If the `isEmptyResponseExpected` is true, then this is not really an error case from the point of the caller. I would expect that it would return successful promise resolution with undefined/void value.

Reply:
> If the isEmptyResponseExpected is true, then this is not really an error case from the point of the caller. I would expect that it would return successful promise resolution with undefined/void value.

Fair point. We can use a more bespoke structure then Either then, with three fields: Left (error), Middle (non-error deviation from main path) Right (expected main path response). If Right is set, we proceed with the main path, if left is set, we handle an error, if Middle is set, we check the code returned and handle it in a custom way.

> The code could check the response status, and if it is 204, then the parsing should be skipped IMO.

Aren't we already doing that?

Reply:
@kibertoad maybe I'm looking at a wrong line, but I think we're not checking the status before parsing. 

![CleanShot 2024-07-08 at 13 33 37@2x](https://github.com/lokalise/frontend-http-client/assets/4136929/3bdc0ee5-3b9b-45bc-9acc-8ce66a1556e6)

Reply:
@ondrejsevcik check the `tryToResolveJsonBody` itself

@CarlosGamero We probably should have the check in the consistent place for both GET and non-GET scenarios, I think currently I did it explicitly for non-GET and implicitly (within `tryToResolveJsonBody`) for GET.

Reply:
@kibertoad I tried to extract the wretch response handling to a separate method and use it everywhere but param types are not compatible so it is not an easy change

Reply:
@kibertoad I will use the same approach on both places 🙇

Reply:
Already changed